### ISSUE TITLE
favicon download does not work anymore

### ISSUE THREAD

**Issue by micktg** [keeweb/keeweb#1489]
**Describe the bug**
The download of the favicon does not work anymore as it used to be.

**Version**
Desktop App v1.14.0

**OS**
Windows 10 and/or macOS Catalina (both the same behavior)

**To Reproduce**
1. Go to any entry with a Website link, e.g. https://www.any.do
2. Click in the top right corner (default icon)
<img width="72" alt="icon1" src="https://user-images.githubusercontent.com/278107/80793293-0812e500-8b97-11ea-92b2-b4dff6629dc9.png">

3. Click on the cloud icon with the down arrow to download the favicon for the current entry
<img width="35" alt="icon2" src="https://user-images.githubusercontent.com/278107/80793294-0812e500-8b97-11ea-8e0c-1b271d3a958b.png">

4. It always appears the "not available sign" even though the URL is correct and shows a favicon in the browser.
<img width="28" alt="icon3" src="https://user-images.githubusercontent.com/278107/80793292-077a4e80-8b97-11ea-803a-72ad995bd184.png">

This feature used to work in the past. Unfortunately I cannot say exactly when, but if I should guess about a half a year ago it stopped working.

**Comment by antelle** [comment]
Hi! Are you using a webapp or desktop app? Which version? Could you also please give a website where it fails?

**Comment by micktg** [comment]
> Hi! Are you using a webapp or desktop app? Which version? Could you also please give a website where it fails?

I updated my original post.
I added an example website but in my case it every entry with a website. It just doesn't work anymore. Tested on Windows 10 and macOS Catalina with the desktop Version. Existing icons, that are previously downloaded are shown, but new ones cannot be downloaded. Also existing ones cannot be overwritten (cause no new icon can be found).
All Firewalls are disabled or the rules allow keeweeb.exe or keeweeb.app explicitly. The Internet connection works, because I can update the app and I have connection to my cloud provider, so that should not be the problem.

**Comment by antelle** [comment]
Could you please try it on https://apple.com (first entry in the demo db)?

**Comment by antelle** [comment]
I see, it's broken on subdomains. Fixed it

**Comment by micktg** [comment]
Not its working. I seriously started to doubt myself for a moment before I noticed your last comment ;)
Thanks for fixing it!

**Comment by micktg** [comment]
> I see, it's broken on subdomains. Fixed it, now it should work.

But some are still not working, for example: https://cloud.docker.com/

**Comment by antelle** [comment]
That one is a bit special because it returns favicon with `application/octet-stream` content-type, and the page itself redirects to hub.docker.com that doesn't have a favicon there. Fixed it, now it works too.

**Comment by micktg** [comment]
> That one is a bit special because it returns favicon with `application/octet-stream` content-type, and the page itself redirects to hub.docker.com that doesn't have a favicon there. Fixed it, now it works too.

Thanks for the great work! I really appreciate it!

If there is nothing more important to do right now, one could take a look why the following favicons do not work yet.

https://keeweb.info/
https://www.fressnapf.de/
https://pay.google.com/
https://komene.com/
https://revivogen.com/
https://www.rmv.de/
https://www.smartversity.de/
https://socialclub.rockstargames.com/
https://www.synology.com/
https://www.tailorbrands.com/

Thanks in advance!

**Comment by antelle** [comment]
Most of services return favicon in a weird way (all of them are different), I'm not sure if it will be possible to fix all of them.

**Comment by micktg** [comment]
> Most of services here return favicons in a weird way (all of them are different), I'm not sure if it will be possible to fix all of them.

Ok, I see. 
If I'm the only one who is bothered by this, it doesn't seem to be such an important feature for most users. Personally, I do like having an icon for an entry, but it is not absolutely necessary.

**Comment by Aetherinox** [comment]
@mickdewald With the recent changes to our Favicon system; I've tested all of the websites you listed above, and they all return favicons now.

A few of them have crappy ones, (like [rmv.de/](https://www.rmv.de/)), but that's just the favicon they are serving.

If you find any sites that have bad favicons and there are better options out there; feel free to submit a replacement at [rmv.de/](https://www.rmv.de/) by submitting a pull request.

If you have any future issues, please submit another issue.

**Comment by mickdewald** [comment]
Oh wow,
Has the repo been revived? I thought this wonderful project was dead.
It would be wonderful if KeeWeb was further developed.

There are still a lot of Favicons, that cannot be fetched.
I can create a list.

**Comment by antelle** [comment]
I had an idea or making this before by the way, what stopped me is thoughts about copyright. Someone may say that the icon is used without a permission. Perhaps just adding a note how to request icon takedown would solve this.

**Comment by Aetherinox** [comment]
@antelle A takedown policy on the repo you mean?

**Comment by antelle** [comment]
Yeah in the favicon-cdn, and maybe to the plaintext index page that is returned from the worker.

**Comment by Aetherinox** [comment]
@antelle Roger, I can do that real quick.

Side note, do we happen to have any type of mailserver for the *keeweb.info domain? Would be nice to have a way to set up SMTP that we can monitor for something along the lines of requests / services [at] keeweb

**Comment by antelle** [comment]
That’s exactly what I was thinking about. Now we have two things:
- my email in the app settings page: needs to be replaced with keeweb@keeweb.info
- keeweb@keeweb.info routed to my email in cloudflare: needs to be something collaborative because support requests are coming there

What do you think?

**Comment by Aetherinox** [comment]
Yeah that addr works for me. I can change the app references. I figured an email for takedowns would be better, considering they're being submitted by companies, they would contain personal information; and I doubt they're going to submit an issue for that.

**Comment by antelle** [comment]
It needs to be an email of course. I’ll try to find some collaborative email service that we can use to route emails to.

**Comment by Aetherinox** [comment]
I've got to read up on it more, but possibly: https://www.cloudflare.com/developer-platform/email-routing/

**Comment by antelle** [comment]
It’s already in use for keeweb@keeweb.info, but we need to route it somewhere to a shared inbox

**Comment by Aetherinox** [comment]
Ah ok, yeah if you can find a service. From what I was reading, by using a worker, you can route the same email address to multiple people. The only downside to that, is it would go to both of our personal inboxes, but we'd have no way to know if either replied.

**Comment by antelle** [comment]
Exactly, it can be some ui where we can see. I used Zoho before CF introduced the routing feature, will check if it supports shared access.

**Comment by antelle** [comment]
Oh it’s free up to 5 users, should be fine

**Comment by mickdewald** [comment]
> @mickdewald Send me a list of some of the ones you're having issues with and I'll look. Try to keep the list to like less than 10. Chances are, if a few are messed up, it'll correct others. It's weird you found some, I've tested maybe 40+ on desktop, local, and webapp, and never seen one not post.
> 
> And also which method you're using for KeeWeb (desktop, web, etc).

The list is tiny, it does actually work for most of them now, great job!

- https://app.notesnook.com/
- https://auth0.auth0.com/
- https://console.aws.amazon.com/console/home
- https://app.docusign.com/

So my guess is, that it has something do to when a URL has a subdomain.

**Comment by Aetherinox** [comment]
Glad to hear a majority are working.
And yeah, you're right, for some reason these are the first ones I've seen not actually fetch the favicon.

Let me go dig into the code, view the page sources, and see what's going on. I was feeling pretty damn happy that I nabbed every one of them up until this point lol

**Comment by Aetherinox** [comment]
@mickdewald You really had to give me a pain the rear problem.

I'm almost done. The big one was actually docusign. I got the others to work. But docusign does define an SVG within its source, which is great. However, it did not define a width and height to the svg. 

So KeeWeb was fetching the svg, but it was appearing as a broken image.

![qggkfdsibT](https://github.com/keeweb/keeweb/assets/118329232/8338a57d-17e7-4498-aa38-aaf4c1a04f27)

So the solution was to check the request as it comes in and see if it has a size defined, and add one if not.

![zf0cWsHLhS](https://github.com/keeweb/keeweb/assets/118329232/e6c0337d-85a4-4f37-8449-ca9ad5ce4ae7)

**Comment by Aetherinox** [comment]
Alright, after those tweaks, I went on a Google rampage. I went down a list of random sites, and grabbed every one.

![vXm2m1oAkX](https://github.com/keeweb/keeweb/assets/118329232/7931887a-de7b-4639-89d4-f2be4682b289)

I used links with all different characters in it, 6 subdomains deep, with https, without, www, and any other type of URL I could find. Every single one of them rendered with an icon.

**Comment by Aetherinox** [comment]
@antelle Also, notice added https://services.keeweb.info/favicon

Hopefully for now, this is done.

**Comment by antelle** [comment]
@Aetherinox added you to a google group

**Comment by mickdewald** [comment]
@Aetherinox Great job, and thanks for the effort!
Sorry, this did so much pain, I guess there are more important things to do in a password manager.

**Comment by Aetherinox** [comment]
@mickdewald Eh no worries. It was a pain at times, but it was also partly due to us moving how we hosted it. Cloudflare workers don't operate in the same way that a typical node project would. They have their own library you're allowed to use, and they don't play with polyfills.

But now it should be able to capture at least 95% of all icons without issue. I added so many damn backup methods, there's like 7 things it tries to find the website's icon before it finally just gives you a default icon.
